Source/JavaScriptCore/b3/B3ReduceStrength.cpp

```diff
@@ -3405,6 +3411,110 @@ class ReduceStrength {
+        if ((m_value->child(0)->opcode() == Shl
+                && m_value->child(1)->opcode() == ZShr)
+            || (m_value->child(0)->opcode() == ZShr
+                && m_value->child(1)->opcode() == Shl)) {
+            Value* shl = m_value->child(0)->opcode() == Shl ? m_value->child(0) : m_value->child(1);
+            Value* shr = m_value->child(0)->opcode() == ZShr ? m_value->child(0) : m_value->child(1);
+
+            // Handle the cases where both shifts have a constant shift amount.
+            if (m_value->type() == Int32
+                && shl->child(1)->hasInt()
+                && shr->child(1)->hasInt()
+                && shl->child(1)->asInt() + shr->child(1)->asInt() == 32
+                && shl->child(0) == shr->child(0)) {
+                replaceWithNew<Value>(RotR, m_value->origin(), shr->child(0), shr->child(1));
+                return true;
+            }
+
+            if (m_value->type() == Int64
+                && shl->child(1)->hasInt()
+                && shr->child(1)->hasInt()
+                && shl->child(1)->asInt() + shr->child(1)->asInt() == 64
+                && shl->child(0) == shr->child(0)) {
+                replaceWithNew<Value>(RotR, m_value->origin(), shr->child(0), shr->child(1));
+                return true;
+            }
+
+            // Otherwise, see if one shift has a Sub operand. We also accept a Sub
+            // that is the single argument of a Trunc node - because this operand is
+            // the shift amount, it semantically acts masked anyway, so the observed
+            // behavior shouldn't change when truncated. This helps when trying to
+            // match this pattern against 64-bit expressions.
+            Value* sub = nullptr;
+            bool isRotL = false;
+            bool isRotR = false;
+            bool needsTrunc = false;
+            if (shl->child(1)->opcode() == Sub) {
+                isRotR = true;
+                sub = shl->child(1);
+            } else if (shl->child(1)->opcode() == Trunc && shl->child(1)->child(0)->opcode() == Sub) {
+                isRotR = true;
+                sub = shl->child(1)->child(0);
+                needsTrunc = true;
+            }
+            if (shr->child(1)->opcode() == Sub) {
+                isRotL = true;
+                sub = shr->child(1);
+            } else if (shr->child(1)->opcode() == Trunc && shr->child(1)->child(0)->opcode() == Sub) {
+                isRotL = true;
+                sub = shr->child(1)->child(0);
+                needsTrunc = true;
+            }
+
+            // Continue only if we have a Sub in exactly one of the two shifts, and
+            // it is subtracting from 32 or 64 respectively based on the type of
+            // this instruction.
+            if (isRotR != isRotL
+                && ((m_value->type() == Int64 && sub->child(0)->isInt(64))
+                    || (m_value->type() == Int32 && sub->child(0)->isInt(32)))) {
+                Value* x = shl->child(0);
```

Comment:
nit: Maybe we can use `x1` for `x` for readability since it aligns to to the comment above.

Reply:
Ah, sure, that comment is newer than this code. I'll change it in the next push.